huntsman/observatory.py

```diff
@@ -232,202 +235,82 @@ def autofocus_cameras(self, *args, **kwargs):
-                camera_event = camera.take_observation(
-                    flat_obs,
-                    fits_headers,
-                    filename=filename,
-                    exptime=exptime,
-                    dark=True
-                )
-
-                camera_events[cam_name] = {
-                    'event': camera_event,
-                    'filename': filename,
-                }
-
-            # Will block here until done exposing on all cameras
-            while not all([info['event'].is_set() for info in camera_events.values()]):
-                self.logger.debug('Waiting for dark-field image')
-                time.sleep(1)
-
-    def take_morning_flats(*args, **kwargs):
-        '''
+        if camera_names is None:
+            cameras_all = self.cameras
+        else:
+            cameras_all = {c: self.cameras[c] for c in camera_names}
+
+        # Obtain the filter order
+        filter_order = self.config['filterwheel']['flat_field_order'].copy()
+        if self.past_midnight:  # If it's the morning, order is reversed
+            filter_order.reverse()
+
+        exptimes_dark = {c: set() for c in cameras_all.keys()}
+        for filter_name in filter_order:
+
+            if not safety_func():
+                self.logger.info('Terminating flat-fielding because its no longer safe.')
+                return
+
+            # Get a dict of cameras that have this filter
+            filter_cameras = {}
+            for cam_name, cam in cameras_all.items():
+                if cam.filterwheel is None:
+                    if cam.filter_type == filter_name:
+                        filter_cameras[cam_name] = cam
+                elif filter_name in cam.filterwheel.filter_names:
+                    filter_cameras[cam_name] = cam
+
+            # Go to next filter if there are no cameras with this one
+            if not filter_cameras:
+                self.logger.debug(f'No cameras found with {filter_name} filter.')
+                continue
+
+            # Create the Observation object
+            obs = self._create_flat_field_observation(alt=alt, az=az, filter_name=filter_name)
+
+            # Take the flats for each camera in this filter
+            self.logger.info(f'Taking flat fields in {filter_name} filter.')
+            exptimes = self._take_autoflats(filter_cameras, obs, safety_func=safety_func, **kwargs)
+
+            # Log the new exposure times we need to take darks with
+            for cam_name in filter_cameras.keys():
+                exptimes_dark[cam_name].update(exptimes[cam_name])
+
+        # Take darks for each exposure time we used
+        if safety_func():
+            self.logger.info('Taking flat field dark frames.')
+            obs = self._create_flat_field_observation(alt=alt, az=az)
+            for exptime in exptimes_dark:
+                self._take_flat_field_darks(exptimes_dark, obs)
+        else:
+            self.logger.info('Skipping flat-field darks as no longer safe.')
```

Comment:
Need to check safety between every iteration.

```suggestion
        obs = self._create_flat_field_observation(alt=alt, az=az)
        for i, exptime in enumerate(exptimes_dark):
            if safety_func():
                self.logger.info('Taking flat field dark frame {i} of {len(exptimes_dark)}.')
                self._take_flat_field_darks(exptimes_dark, obs)
            else:
                self.logger.info('Stopping flat-field darks as no longer safe.')
                break
```